### PR TITLE
Enable group sync for LDAP servers, that do not return memberOf

### PR DESCRIPTION
**Pull request by rdd2**
The OpenLDAP setup we use does not return functional attributes by default. I understand this behaviour is quite common. One has to query for memberOf separately. So to enable allocating users to groups, I have added an extra ldap_read() that adds memberOf attributes to the lastAuthProperties array. 

This is relevant to issue #4454.

### CODE REVIEW

**Comment by AdSchellevis** [comment]
it's probably better to optionally set the attributes to return with an additional parameter, this duplicates most of the code.

**Comment by rdd2** [comment]
I agree that that would be the superior way of doing it, but I don't think it would work. From the [ldapsearch manpage](http://manpages.ubuntu.com/manpages/impish/man1/ldapsearch.1.html):

> If ldapsearch finds one or more entries, the attributes specified by attrs are returned. If * is listed, all user attributes are returned. If + is listed, all operational attributes are returned. If no attrs are listed, all user attributes are returned. If only 1.1 is listed, no attributes will be returned. 

Note that it makes no mention of combining the two wildcards. And from what I could test with a few (Open)LDAP-Servers this is indeed the case. You can query ```+``` and get functional attributes or ```*``` (or empty attribute list) and get user attributes, but you cannot combine them to get all attributes. I'd be interested if someone could test with other LDAP servers how they behave.

It is of course possible to list all attributes by listing them all and tacking ```memberof``` onto the end, but that is not really a good way here. 

Applying this to the situation at hand we definitely need two queries for this type of server. So I considered artificially joining the two result arrays but that would have been a bit clumsy with the ```count``` key and all. 

So I opted to do it this way, to run a separate query in exactly this situation and process the second query exactly like the first one. If you want I could do a slightly more elegant version where I add both results to an array and loop through them with the processing block. I however prefer the existing commit for readability.

Thanks for getting into this so quickly.

Thomas

**Comment by AdSchellevis** [comment]
Hi Thomas,

If I'm not mistaken, the only change in your code is the following line:

```
$sr = @ldap_read($this->ldapHandle, $user_dn, '(objectclass=*)');
$sr_memberOf = @ldap_read($this->ldapHandle, $user_dn, '(objectclass=*)', array("memberOf"));
```

Which seems to suggest that `['memberOf']` could be an optional list of parameters to return, in which case an additional parameter for only this option would be enough to reach your goal, wouldn't break existing setups and doesn't query twice when it doesn't have to.

Best regards,

Ad

**Comment by rdd2** [comment]
Hi Ad, 

yes, this is the change. The thing is, that the second query will ONLY return the ```memberOf``` attribute and value(s) and not all the other attributes. The first query is an implicit wildcard query because it names no attributes. It will only return user attributes (and not ```memberOf``` because it is a operational attribute). AFAIK it is not possibe to combine the wildcards with other attributes.

I my first approach to this was the same but that does not work. 

Best regards

Thomas

P.S. I wrote "functional" instead of operational earlier. Just corrected that.

**Comment by AdSchellevis** [comment]
Hi Stephan,  

Just read my previous comment again and keep in mind that there's no need to do both (fetch all or define what you need).

Best regards,

Ad

**Comment by fichtner** [comment]
> AFAIK it is not possibe to combine the wildcards with other attributes.

To query e.g. hidden attributes createTimestamp and modifyTimestamp my query is [ '*', 'createTimestamp', 'modifyTimestamp' ] and I have to say it works. :)


Cheers,
Franco

**Comment by rdd2** [comment]
Hi Franco, 

you are right. That does work. I just tried. That is of course a much better way of doing it. I'll be back with a new commit. 

Thank you. 

Thomas

**Comment by fichtner** [comment]
wouldn't it be better to pull it unconditionally. what's the worst that could happen?

**Comment by rdd2** [comment]
Since there are LDAP servers which do not support memberOf at all and I can't be sure how they react when queried explicitly, I would make it conditional on users actually activating group sync.

**Comment by fichtner** [comment]
Then you have the same problem on a smaller user base. It just got a lot harder to prove your assumption.

**Comment by rdd2** [comment]
> Then you have the same problem on a smaller user base. It just got a lot harder to prove your assumption.
Please explain. 

As I see it, there are 3 scenarios:
- LDAP server with no memberOf overlay/functionality -> will work for user auth but will not for group sync and may(!) throw an error when queried for memberOf. This can then be remedied by deselecting group sync.
- LDAP with memberOf which answers with memberOf on empty attribute list -> worked before and will work with this patch
- LDAP with memberOf which only returns user attributes on empty attribute list -> did not work before but works with this patch

**Comment by fichtner** [comment]
There is an easy way to test how a server responds to non-existing attributes. My guess is that it will not complain at all.

**Comment by rdd2** [comment]
> There is an easy way to test how a server responds to non-existing attributes. My guess is that it will not complain at all.

Yes, that is also my experience. I have tested with OpenLDAP and Samba4. Samba4 includes memberOf in its wildcard answer and also does not complain on non-existing attributes. I would bet this this also the way AD behaves. There are however quite a few LDAP implementations and I don't have access to them all. If you want I could just add the attributes to the standard query.

File: src/opnsense/mvc/app/library/OPNsense/Auth/LDAP.php [hunk 1]
```diff
@@ -578,7 +578,7 @@ public function authenticate($username, $password)
         if ($ldap_is_connected) {
             $this->lastAuthProperties['dn'] = $user_dn;
             if ($this->ldapReadProperties) {
-                $sr = @ldap_read($this->ldapHandle, $user_dn, '(objectclass=*)');
+                $sr = @ldap_read($this->ldapHandle, $user_dn, '(objectclass=*)', array("*", "memberOf"));
```
Comment (fichtner) [inline src/opnsense/mvc/app/library/OPNsense/Auth/LDAP.php:581]:
```suggestion
                $sr = @ldap_read($this->ldapHandle, $user_dn, '(objectclass=*)', ['*', 'memberOf']);
```

**Review by fichtner — APPROVED** [review]
neat and clean, very nice

**Comment by rdd2** [comment]
Great thank you for all the feedback. This is my first PR in OPNsense. Yay!

**Comment by fichtner** [comment]
Thanks for bearing with us, merged! ❤️

**Pull request merged**